sound/soc/sof/pm.c

```diff
@@ -260,15 +260,61 @@ static void sof_cache_debugfs(struct snd_sof_dev *sdev)
 }
 #endif
 
+/*
+ * DSP D0/D3 States may transform as below:-
+ *
+ *       -------------------
+ *       |            d0i0 |
+ *       |       D0        |<--------------+
+ *       |            d0i3 |               |
+ *       -------------------               |
+ *          ^          |                   |
+ *          |          |suspend            |
+ *          |          v                   |
+ *          |  -------------------         |
+ *   resume |  | kwd_at_s0i3 &&  |   Yes   |
+ *          |  | target_is_s0i3  |---------+
+ *          |  |                 |
+ *          |  -------------------
+ *          |          |
+ *          |          | No
+ *          |          v
+ *       -------------------
+ *       |                 |
+ *       |       D3        |
+ *       |                 |
+ *       -------------------
+ *
+ */
+
 static int sof_resume(struct device *dev, bool runtime_resume)
 {
 	struct snd_sof_dev *sdev = dev_get_drvdata(dev);
+	struct pci_dev *pci = to_pci_dev(sdev->dev);
 	int ret;
 
 	/* do nothing if dsp resume callbacks are not set */
 	if (!sof_ops(sdev)->resume || !sof_ops(sdev)->runtime_resume)
 		return 0;
 
+	if (sdev->kwd_at_s0i3 &&
+	    acpi_target_system_state() == ACPI_STATE_S0) {
```

Comment:
so what you are trying to say is that the system is trying to enter S0ix while the device is in D0ix. You absolutely don't need this 'kdw_at_s0i3' kludge.

Reply:
maybe the naming here is confusion, any suggestion?

Reply:
no it's not a naming issue. it's the fact that you already have a substate, so use it!